src/test/java/io/vertx/rabbitmq/RabbitMQServiceTest.java

```diff
@@ -1,33 +1,82 @@
 package io.vertx.rabbitmq;
 
-import static io.vertx.test.core.TestUtils.randomAlphaString;
-import static io.vertx.test.core.TestUtils.randomInt;
-
 import com.rabbitmq.client.AMQP;
-import com.rabbitmq.client.Channel;
-import com.rabbitmq.client.ConnectionFactory;
 import io.vertx.core.Handler;
 import io.vertx.core.json.JsonObject;
-import io.vertx.test.core.VertxTestBase;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.rabbitmq.impl.RabbitMQClientImpl;
 import org.junit.Test;
 
+import java.io.IOException;
 import java.util.HashMap;
 import java.util.HashSet;
+import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ArrayBlockingQueue;
 import java.util.concurrent.CountDownLatch;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static io.vertx.test.core.TestUtils.randomAlphaString;
+import static io.vertx.test.core.TestUtils.randomInt;
 
 /**
  * @author <a href="mailto:nscavell@redhat.com">Nick Scavelli</a>
  */
 public class RabbitMQServiceTest extends RabbitMQClientTestBase {
 
+  private static final Logger log = LoggerFactory.getLogger(RabbitMQServiceTest.class);
+
   @Override
   public void setUp() throws Exception {
     super.setUp();
     connect();
   }
 
+  @Test
+  public void testMessageOrderingWhenConsuming() throws IOException, InterruptedException {
+
+    String queueName = "message_ordering_test";
+    String address = queueName + ".address";
+
+    int count = 1000;
+
+    CountDownLatch latch = new CountDownLatch(count - 1);
+    List<String> sendingOrder = IntStream.range(1, count).boxed().map(Object::toString).collect(Collectors.toList());
+
+    ArrayBlockingQueue<String> receiveOrderQueue = new ArrayBlockingQueue<>(count);
+    receiveOrderQueue.addAll(sendingOrder);
+
+    // set up queue
+    AMQP.Queue.DeclareOk ok = channel.queueDeclare(queueName, false, false, true, null);
+    assertNotNull(ok.getQueue());
+    AMQP.BasicProperties properties = new AMQP.BasicProperties.Builder().contentType("text/plain").contentEncoding("UTF-8").build();
+
+    // send  messages
+    for (String msg : sendingOrder)
+      channel.basicPublish("", queueName, properties, msg.getBytes("UTF-8"));
+
+    vertx.eventBus().consumer(address, msg -> {
+      String expectedMessage = receiveOrderQueue.poll();
+      assertNotNull(msg);
+      JsonObject json = (JsonObject) msg.body();
+      assertNotNull(json);
+      String body = json.getString("body");
+      assertNotNull(body);
+      log.info("received: " + body + " expected: " + expectedMessage);
+      assertTrue(body.equals(expectedMessage));
+      latch.countDown();
+    });
+
+    client.basicConsume(queueName, address, onSuccess(v -> {
+    }));
+
+    awaitLatch(latch);
+    testComplete();
```

Comment:
testComplete should not be called here, it is only necessary when called asynchronously and if you call `await()` in the junit thread

Reply:
removed